### PR TITLE
metrics: add col/idx name(s) for BackfillProgressGauge and BackfillTotalCounter

### LINKED ISSUE #58114: Include index name in `tidb_ddl_add_index_total` metrics
## Enhancement

This is an improvement in the observability of adding index DDL.

### PR DESCRIPTION
**Pull request by CbcWestwolf**
### What problem does this PR solve?

Issue Number: close #58114

Problem Summary:

For BackfillProgressGauge and BackfillTotalCounter, tidb only shows the database name and table name in the metrics. If we add multiple indexes on a single table, it would be difficult to know the rate and progress of each index.

### What changed and how does it work?

* Add column names and index names for metrics.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Test script:
```bash
mysql -h 127.0.0.1 -P4000 -uroot -e'create database test1; create database test2;'

sysbench oltp_read_write prepare --db-driver=mysql --mysql-host=127.0.0.1 --mysql-port=4000 --mysql-user=root --mysql-db=test1 --threads=16 --tables=2 --table-size=1000000
sysbench oltp_read_write prepare --db-driver=mysql --mysql-host=127.0.0.1 --mysql-port=4000 --mysql-user=root --mysql-db=test2 --threads=16 --tables=2 --table-size=1000000

# Add Index
mysql -h 127.0.0.1 -P4000 -uroot -e'create index k_pad on test1.sbtest1(pad); alter table test1.sbtest2 add index k_pad(pad), add index k_comp(c, pad);'
mysql -h 127.0.0.1 -P4000 -uroot -e'create index k_pad on test2.sbtest1(pad); alter table test2.sbtest2 add index k_pad(pad), add index k_comp(c, pad);'

# Modify Column
mysql -h 127.0.0.1 -P4000 -uroot -e'alter table test1.sbtest1 change column k k varchar(64); alter table test1.sbtest2 change column k k varchar(64);'
mysql -h 127.0.0.1 -P4000 -uroot -e'alter table test2.sbtest1 change column k k varchar(64); alter table test2.sbtest2 change column k k varchar(64);'
```

image

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/metrics/ddl.go [hunk 1]
```diff
@@ -180,26 +180,48 @@ func InitDDLMetrics() {
...
 	var stringBuilder strings.Builder
-	stringBuilder.Grow(len(label) + len(schemaName) + len(tableName) + 2)
+	if len(colOrIdxNames) == 0 {
+		stringBuilder.Grow(len(label) + len(schemaName) + len(tableName) + 6)
+	} else {
+		stringBuilder.Grow(len(label) + len(schemaName) + len(tableName) + len(colOrIdxNames) + 9)
+	}
 	stringBuilder.WriteString(label)
-	stringBuilder.WriteString("_")
+	stringBuilder.WriteString("_[")
```
Comment (tangenta) [inline pkg/metrics/ddl.go:207]: Is there another usage for "[" in labels?

**Comment by fzzf678** [comment]
Please consult that what's the unit on the vertical axis in `DDL Backfill Data Rate`, can it be more than 100%?
image

File: pkg/metrics/ddl.go (hunk 1 above)
Comment (CbcWestwolf) [reply pkg/metrics/ddl.go:207]: Not found. Maybe `dbname-tablename-col/idxname` is a better way. Let me modify it

Comment (CbcWestwolf) [reply pkg/metrics/ddl.go]: Fixed in https://github.com/pingcap/tidb/pull/58380/commits/5b8a1117004ce9aba4dfebfc1b53f63a021cc97f, PTAL

**Review by tangenta — APPROVED** [review]

**Review by fzzf678 — APPROVED** [review]

**Pull request merged**